althea_kernel_interface/src/iface_counter.rs

```diff
@@ -23,28 +39,87 @@ impl KernelInterface {
         Ok(())
     }
 
-    /// returns ((input bytes, input packets), (output bytes, output packets))
-    pub fn read_iface_counters(&self, interface: &str) -> Result<((u64, u64), (u64, u64)), Error> {
+    /// returns (input counters, output counters)
+    pub fn read_iface_counters(
+        &self,
+        interface: &str,
+    ) -> Result<(IfaceCounter, IfaceCounter), Error> {
+        lazy_static! {
+            static ref RE: Regex =
+                Regex::new(r"(?m)^\s+(?P<pkts>\d+)\s+(?P<bytes>\d+)\s+all\s+--\s+(?P<in>.+?)\s+(?P<out>.+?)\s+")
+                    .expect("Unable to compile regular expression");
+        }
         let chain_name = format!("{}-counter", interface);
 
         let output = self.run_command("iptables", &["-w", "-L", &chain_name, "-Z", "-x", "-v"])?;
 
         let stdout = String::from_utf8(output.stdout)?;
 
-        let re = Regex::new(&format!(
-            r"(?m)^\s+(\d+)\s+(\d+)\s+all\s+--\s+any\s+{}",
-            interface
-        )).unwrap();
-        let caps = re.captures(&stdout).unwrap();
-        let output_traffic = (caps[2].parse::<u64>()?, caps[1].parse::<u64>()?);
-
-        let re = Regex::new(&format!(
-            r"(?m)^\s+(\d+)\s+(\d+)\s+all\s+--\s+{}\s+any",
-            interface
-        )).unwrap();
-        let caps = re.captures(&stdout).unwrap();
-        let input_traffic = (caps[2].parse::<u64>()?, caps[1].parse::<u64>()?);
-
-        Ok((input_traffic, output_traffic))
+        let mut input: Option<IfaceCounter> = None;
+        let mut output: Option<IfaceCounter> = None;
+
+        for caps in RE.captures_iter(&stdout) {
+            if &caps["in"] == "any" && &caps["out"] == interface {
+                output = Some(IfaceCounter::new(
+                    caps["bytes"].parse()?,
+                    caps["pkts"].parse()?,
+                ));
+            } else if &caps["in"] == interface && &caps["out"] == "any" {
+                input = Some(IfaceCounter::new(
+                    caps["bytes"].parse()?,
+                    caps["pkts"].parse()?,
+                ));
+            }
+        }
+
+        if input.is_none() || output.is_none() {
+            error!("Unable to parse iface counters: {}", stdout);
+        }
+
+        Ok((
```

Comment:
um we're returning Ok(Err)?

Reply:
No - here `.expect` panics with a nice message. It could (and should) be done better with a custom error perhaps to handle it better.